services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusMigrationServiceTest.kt

```diff
@@ -0,0 +1,238 @@
+                    ApplicationFactory.createContact(firstName = "Pekka", lastName = " "),
+                    ApplicationFactory.createContact(firstName = "Matti", email = "", phone = null),
+                    ApplicationFactory.createContact(
+                        firstName = "Hannu",
+                        lastName = null,
+                        email = " "
+                    )
+                )
+
+            val response = contacts.contactWithLeastMissingFields()
+
+            assertThat(response).isNotNull().isSameInstanceAs(contacts[1])
+        }
+    }
+
+    @Nested
+    inner class FindOrderer {
+        @Test
+        fun `returns null when there are no contacts`() {
+            val data =
+                ApplicationFactory.createCableReportApplicationData(
+                    customerWithContacts = null,
+                    contractorWithContacts = null
+                )
+
+            val result = HakemusMigrationService.findOrderer(data)
+
+            assertThat(result).isNull()
+        }
+
+        @Test
+        fun `returns the orderer when there is just one`() {
+            val data =
+                ApplicationFactory.createCableReportApplicationData(
+                    customerWithContacts = ApplicationFactory.createCompanyCustomerWithOrderer(),
+                    contractorWithContacts =
+                        ApplicationFactory.createCompanyCustomer().withContact()
+                )
+
+            val result = HakemusMigrationService.findOrderer(data)
+
+            assertThat(result).isSameInstanceAs(data.customerWithContacts!!.contacts[0])
+        }
+
+        @Test
+        fun `returns the orderer with the least missing fields if there are several`() {
+            val nonOrderer =
+                ApplicationFactory.createContact(
+                    orderer = false,
+                    firstName = null,
+                    lastName = "",
+                    phone = " "
+                )
+            val missingTwo =
+                ApplicationFactory.createContact(
+                    orderer = true,
+                    firstName = "Pekka",
+                    lastName = " ",
+                    email = ""
+                )
+            val firstMissingOne =
+                ApplicationFactory.createContact(orderer = true, firstName = "Matti", phone = null)
+            val secondMissingOne =
+                ApplicationFactory.createContact(orderer = true, firstName = "Hannu", email = " ")
+            val contacts = arrayOf(nonOrderer, missingTwo, firstMissingOne, secondMissingOne)
+            val data =
+                ApplicationFactory.createCableReportApplicationData(
+                    customerWithContacts =
+                        ApplicationFactory.createCompanyCustomer().withContacts(*contacts),
+                )
+
+            val result = HakemusMigrationService.findOrderer(data)
+
+            assertThat(result).isSameInstanceAs(firstMissingOne)
```

Comment:
This is more clear than the previous test with the differences between test objects.

Reply:
I changed the test to create a `Contact` directly, without using the factory method. I think this makes it easier to see which fields are missing and which are not.